// @mw-angular/core/esm5/lib/services/mw-error-lock.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { BehaviorSubject } from 'rxjs';
import * as i0 from "@angular/core";
var MwErrorLockService = /** @class */ (function () {
    function MwErrorLockService() {
        this.errorLockSubject = new BehaviorSubject(null);
    }
    /**
     * @return {?}
     */
    MwErrorLockService.prototype.getErrorLock = /**
     * @return {?}
     */
    function () {
        return this.errorLockSubject.asObservable();
    };
    /**
     * @param {?} errorLock
     * @return {?}
     */
    MwErrorLockService.prototype.setErrorLock = /**
     * @param {?} errorLock
     * @return {?}
     */
    function (errorLock) {
        this.errorLockSubject.next(errorLock);
    };
    MwErrorLockService.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root',
                },] }
    ];
    /** @nocollapse */ MwErrorLockService.ngInjectableDef = i0.ɵɵdefineInjectable({ factory: function MwErrorLockService_Factory() { return new MwErrorLockService(); }, token: MwErrorLockService, providedIn: "root" });
    return MwErrorLockService;
}());
export { MwErrorLockService };
if (false) {
    /**
     * @type {?}
     * @private
     */
    MwErrorLockService.prototype.errorLockSubject;
}
