// @mw-angular/core/esm5/lib/services/mw-local-storage.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { MwEnvironmentService } from './mw-environment.service';
import { MwPlatformService } from './mw-platform.service';
import * as i0 from "@angular/core";
import * as i1 from "./mw-platform.service";
import * as i2 from "./mw-environment.service";
var MwLocalStorageService = /** @class */ (function () {
    function MwLocalStorageService(mwPlatformService, mwEnvironmentService) {
        this.mwPlatformService = mwPlatformService;
        this.mwEnvironmentService = mwEnvironmentService;
        this.envKeyPrefix = 'localStorageKeyPrefix';
    }
    /**
     * @private
     * @param {?} data
     * @return {?}
     */
    MwLocalStorageService.prototype.serialize = /**
     * @private
     * @param {?} data
     * @return {?}
     */
    function (data) {
        return JSON.stringify(data);
    };
    /**
     * @private
     * @template T
     * @param {?} data
     * @return {?}
     */
    MwLocalStorageService.prototype.unserialize = /**
     * @private
     * @template T
     * @param {?} data
     * @return {?}
     */
    function (data) {
        return JSON.parse(data);
    };
    /**
     * @param {?} key
     * @param {?} value
     * @return {?}
     */
    MwLocalStorageService.prototype.setItem = /**
     * @param {?} key
     * @param {?} value
     * @return {?}
     */
    function (key, value) {
        if (this.mwPlatformService.isBrowser()) {
            /** @type {?} */
            var serializedData = this.serialize(value);
            localStorage.setItem(this.getPrefixedKey(key), serializedData);
        }
    };
    /**
     * @template T
     * @param {?} key
     * @return {?}
     */
    MwLocalStorageService.prototype.getItem = /**
     * @template T
     * @param {?} key
     * @return {?}
     */
    function (key) {
        if (this.mwPlatformService.isBrowser()) {
            /** @type {?} */
            var value = localStorage.getItem(this.getPrefixedKey(key));
            if (value !== null) {
                return this.unserialize(value);
            }
        }
        return null;
    };
    /**
     * @param {?} key
     * @return {?}
     */
    MwLocalStorageService.prototype.deleteItem = /**
     * @param {?} key
     * @return {?}
     */
    function (key) {
        if (this.mwPlatformService.isBrowser()) {
            localStorage.removeItem(this.getPrefixedKey(key));
        }
    };
    /**
     * @return {?}
     */
    MwLocalStorageService.prototype.clear = /**
     * @return {?}
     */
    function () {
        if (this.mwPlatformService.isBrowser()) {
            localStorage.clear();
        }
    };
    /**
     * @private
     * @param {?} key
     * @return {?}
     */
    MwLocalStorageService.prototype.getPrefixedKey = /**
     * @private
     * @param {?} key
     * @return {?}
     */
    function (key) {
        return this.mwEnvironmentService.getValue(this.envKeyPrefix, '') + key;
    };
    MwLocalStorageService.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root',
                },] }
    ];
    /** @nocollapse */
    MwLocalStorageService.ctorParameters = function () { return [
        { type: MwPlatformService },
        { type: MwEnvironmentService }
    ]; };
    /** @nocollapse */ MwLocalStorageService.ngInjectableDef = i0.ɵɵdefineInjectable({ factory: function MwLocalStorageService_Factory() { return new MwLocalStorageService(i0.ɵɵinject(i1.MwPlatformService), i0.ɵɵinject(i2.MwEnvironmentService)); }, token: MwLocalStorageService, providedIn: "root" });
    return MwLocalStorageService;
}());
export { MwLocalStorageService };
if (false) {
    /**
     * @type {?}
     * @private
     */
    MwLocalStorageService.prototype.envKeyPrefix;
    /**
     * @type {?}
     * @private
     */
    MwLocalStorageService.prototype.mwPlatformService;
    /**
     * @type {?}
     * @private
     */
    MwLocalStorageService.prototype.mwEnvironmentService;
}
