// @mw-angular/core/esm5/lib/services/mw-loading.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { BehaviorSubject, Observable } from 'rxjs';
import { debounceTime, distinctUntilChanged, map } from 'rxjs/operators';
import * as i0 from "@angular/core";
var MwLoadingService = /** @class */ (function () {
    function MwLoadingService() {
        this.isLoadingSubjectsPool = {};
    }
    /**
     * @param {?=} tag
     * @return {?}
     */
    MwLoadingService.prototype.getIsLoading = /**
     * @param {?=} tag
     * @return {?}
     */
    function (tag) {
        if (tag === void 0) { tag = 'general'; }
        this.checkAndInitIsLoadingSubject(tag);
        return this.isLoadingSubjectsPool[tag].asObservable().pipe(map((/**
         * @param {?} value
         * @return {?}
         */
        function (value) { return value > 0; })), distinctUntilChanged(), debounceTime(100));
    };
    /**
     * @param {?=} tag
     * @return {?}
     */
    MwLoadingService.prototype.startObservable = /**
     * @param {?=} tag
     * @return {?}
     */
    function (tag) {
        var _this = this;
        if (tag === void 0) { tag = 'general'; }
        return new Observable((/**
         * @param {?} subscriber
         * @return {?}
         */
        function (subscriber) {
            _this.start(tag);
            subscriber.next();
            subscriber.complete();
            return (/**
             * @return {?}
             */
            function () { });
        }));
    };
    /**
     * @param {?=} tag
     * @return {?}
     */
    MwLoadingService.prototype.start = /**
     * @param {?=} tag
     * @return {?}
     */
    function (tag) {
        if (tag === void 0) { tag = 'general'; }
        this.checkAndInitIsLoadingSubject(tag);
        this.isLoadingSubjectsPool[tag].next(this.isLoadingSubjectsPool[tag].value + 1);
    };
    /**
     * @param {?=} tag
     * @return {?}
     */
    MwLoadingService.prototype.stop = /**
     * @param {?=} tag
     * @return {?}
     */
    function (tag) {
        if (tag === void 0) { tag = 'general'; }
        if (this.isLoadingSubjectsPool[tag] === undefined) {
            throw new Error("Loading subject was not created for tag: " + tag + ".");
        }
        if (this.isLoadingSubjectsPool[tag].value > 0) {
            this.isLoadingSubjectsPool[tag].next(this.isLoadingSubjectsPool[tag].value - 1);
        }
    };
    /**
     * @param {?=} tag
     * @return {?}
     */
    MwLoadingService.prototype.destroy = /**
     * @param {?=} tag
     * @return {?}
     */
    function (tag) {
        if (tag === void 0) { tag = 'general'; }
        if (this.isLoadingSubjectsPool[tag] !== undefined) {
            this.isLoadingSubjectsPool[tag].next(0);
            this.isLoadingSubjectsPool[tag].complete();
            delete this.isLoadingSubjectsPool[tag];
        }
    };
    /**
     * @private
     * @param {?} tag
     * @return {?}
     */
    MwLoadingService.prototype.checkAndInitIsLoadingSubject = /**
     * @private
     * @param {?} tag
     * @return {?}
     */
    function (tag) {
        if (this.isLoadingSubjectsPool[tag] === undefined) {
            this.isLoadingSubjectsPool[tag] = new BehaviorSubject(0);
        }
    };
    MwLoadingService.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root',
                },] }
    ];
    /** @nocollapse */ MwLoadingService.ngInjectableDef = i0.ɵɵdefineInjectable({ factory: function MwLoadingService_Factory() { return new MwLoadingService(); }, token: MwLoadingService, providedIn: "root" });
    return MwLoadingService;
}());
export { MwLoadingService };
if (false) {
    /**
     * @type {?}
     * @private
     */
    MwLoadingService.prototype.isLoadingSubjectsPool;
}
