// @mw-angular/core/esm5/lib/services/mw-platform.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { isPlatformBrowser, isPlatformServer } from '@angular/common';
import { Inject, Injectable, PLATFORM_ID } from '@angular/core';
import * as i0 from "@angular/core";
var MwPlatformService = /** @class */ (function () {
    function MwPlatformService(platformId) {
        this.platformId = platformId;
    }
    /**
     * @return {?}
     */
    MwPlatformService.prototype.isBrowser = /**
     * @return {?}
     */
    function () {
        return isPlatformBrowser(this.platformId);
    };
    /**
     * @return {?}
     */
    MwPlatformService.prototype.isServer = /**
     * @return {?}
     */
    function () {
        return isPlatformServer(this.platformId);
    };
    MwPlatformService.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root',
                },] }
    ];
    /** @nocollapse */
    MwPlatformService.ctorParameters = function () { return [
        { type: undefined, decorators: [{ type: Inject, args: [PLATFORM_ID,] }] }
    ]; };
    /** @nocollapse */ MwPlatformService.ngInjectableDef = i0.ɵɵdefineInjectable({ factory: function MwPlatformService_Factory() { return new MwPlatformService(i0.ɵɵinject(i0.PLATFORM_ID)); }, token: MwPlatformService, providedIn: "root" });
    return MwPlatformService;
}());
export { MwPlatformService };
if (false) {
    /**
     * @type {?}
     * @private
     */
    MwPlatformService.prototype.platformId;
}
