// @mw-angular/core/esm2015/lib/services/mw-error-lock.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { BehaviorSubject } from 'rxjs';
import * as i0 from "@angular/core";
export class MwErrorLockService {
    constructor() {
        this.errorLockSubject = new BehaviorSubject(null);
    }
    /**
     * @return {?}
     */
    getErrorLock() {
        return this.errorLockSubject.asObservable();
    }
    /**
     * @param {?} errorLock
     * @return {?}
     */
    setErrorLock(errorLock) {
        this.errorLockSubject.next(errorLock);
    }
}
MwErrorLockService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root',
            },] }
];
/** @nocollapse */ MwErrorLockService.ngInjectableDef = i0.ɵɵdefineInjectable({ factory: function MwErrorLockService_Factory() { return new MwErrorLockService(); }, token: MwErrorLockService, providedIn: "root" });
if (false) {
    /**
     * @type {?}
     * @private
     */
    MwErrorLockService.prototype.errorLockSubject;
}
