// @mw-angular/core/esm2015/lib/services/mw-html.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Inject, Injectable, RendererFactory2 } from '@angular/core';
import { DOCUMENT } from '@angular/common';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common";
export class MwHtmlService {
    /**
     * @param {?} rendererFactory
     * @param {?} document
     */
    constructor(rendererFactory, document) {
        this.document = document;
        this.renderer = rendererFactory.createRenderer(null, null);
    }
    /**
     * @param {?} selector
     * @param {?} attributeName
     * @param {?} attributeValue
     * @return {?}
     */
    updateTags(selector, attributeName, attributeValue) {
        Array.from(this.document.querySelectorAll(selector)).forEach((/**
         * @param {?} element
         * @return {?}
         */
        (element) => {
            this.renderer.setAttribute(element, attributeName, attributeValue);
        }));
    }
}
MwHtmlService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root',
            },] }
];
/** @nocollapse */
MwHtmlService.ctorParameters = () => [
    { type: RendererFactory2 },
    { type: undefined, decorators: [{ type: Inject, args: [DOCUMENT,] }] }
];
/** @nocollapse */ MwHtmlService.ngInjectableDef = i0.ɵɵdefineInjectable({ factory: function MwHtmlService_Factory() { return new MwHtmlService(i0.ɵɵinject(i0.RendererFactory2), i0.ɵɵinject(i1.DOCUMENT)); }, token: MwHtmlService, providedIn: "root" });
if (false) {
    /**
     * @type {?}
     * @private
     */
    MwHtmlService.prototype.renderer;
    /**
     * @type {?}
     * @private
     */
    MwHtmlService.prototype.document;
}
