// @mw-angular/core/esm2015/lib/services/mw-local-storage.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { MwEnvironmentService } from './mw-environment.service';
import { MwPlatformService } from './mw-platform.service';
import * as i0 from "@angular/core";
import * as i1 from "./mw-platform.service";
import * as i2 from "./mw-environment.service";
export class MwLocalStorageService {
    /**
     * @param {?} mwPlatformService
     * @param {?} mwEnvironmentService
     */
    constructor(mwPlatformService, mwEnvironmentService) {
        this.mwPlatformService = mwPlatformService;
        this.mwEnvironmentService = mwEnvironmentService;
        this.envKeyPrefix = 'localStorageKeyPrefix';
    }
    /**
     * @private
     * @param {?} data
     * @return {?}
     */
    serialize(data) {
        return JSON.stringify(data);
    }
    /**
     * @private
     * @template T
     * @param {?} data
     * @return {?}
     */
    unserialize(data) {
        return JSON.parse(data);
    }
    /**
     * @param {?} key
     * @param {?} value
     * @return {?}
     */
    setItem(key, value) {
        if (this.mwPlatformService.isBrowser()) {
            /** @type {?} */
            const serializedData = this.serialize(value);
            localStorage.setItem(this.getPrefixedKey(key), serializedData);
        }
    }
    /**
     * @template T
     * @param {?} key
     * @return {?}
     */
    getItem(key) {
        if (this.mwPlatformService.isBrowser()) {
            /** @type {?} */
            const value = localStorage.getItem(this.getPrefixedKey(key));
            if (value !== null) {
                return this.unserialize(value);
            }
        }
        return null;
    }
    /**
     * @param {?} key
     * @return {?}
     */
    deleteItem(key) {
        if (this.mwPlatformService.isBrowser()) {
            localStorage.removeItem(this.getPrefixedKey(key));
        }
    }
    /**
     * @return {?}
     */
    clear() {
        if (this.mwPlatformService.isBrowser()) {
            localStorage.clear();
        }
    }
    /**
     * @private
     * @param {?} key
     * @return {?}
     */
    getPrefixedKey(key) {
        return this.mwEnvironmentService.getValue(this.envKeyPrefix, '') + key;
    }
}
MwLocalStorageService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root',
            },] }
];
/** @nocollapse */
MwLocalStorageService.ctorParameters = () => [
    { type: MwPlatformService },
    { type: MwEnvironmentService }
];
/** @nocollapse */ MwLocalStorageService.ngInjectableDef = i0.ɵɵdefineInjectable({ factory: function MwLocalStorageService_Factory() { return new MwLocalStorageService(i0.ɵɵinject(i1.MwPlatformService), i0.ɵɵinject(i2.MwEnvironmentService)); }, token: MwLocalStorageService, providedIn: "root" });
if (false) {
    /**
     * @type {?}
     * @private
     */
    MwLocalStorageService.prototype.envKeyPrefix;
    /**
     * @type {?}
     * @private
     */
    MwLocalStorageService.prototype.mwPlatformService;
    /**
     * @type {?}
     * @private
     */
    MwLocalStorageService.prototype.mwEnvironmentService;
}
