// @mw-angular/core/esm2015/lib/services/mw-loading.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { BehaviorSubject, Observable } from 'rxjs';
import { debounceTime, distinctUntilChanged, map } from 'rxjs/operators';
import * as i0 from "@angular/core";
export class MwLoadingService {
    constructor() {
        this.isLoadingSubjectsPool = {};
    }
    /**
     * @param {?=} tag
     * @return {?}
     */
    getIsLoading(tag = 'general') {
        this.checkAndInitIsLoadingSubject(tag);
        return this.isLoadingSubjectsPool[tag].asObservable().pipe(map((/**
         * @param {?} value
         * @return {?}
         */
        (value) => value > 0)), distinctUntilChanged(), debounceTime(100));
    }
    /**
     * @param {?=} tag
     * @return {?}
     */
    startObservable(tag = 'general') {
        return new Observable((/**
         * @param {?} subscriber
         * @return {?}
         */
        (subscriber) => {
            this.start(tag);
            subscriber.next();
            subscriber.complete();
            return (/**
             * @return {?}
             */
            () => { });
        }));
    }
    /**
     * @param {?=} tag
     * @return {?}
     */
    start(tag = 'general') {
        this.checkAndInitIsLoadingSubject(tag);
        this.isLoadingSubjectsPool[tag].next(this.isLoadingSubjectsPool[tag].value + 1);
    }
    /**
     * @param {?=} tag
     * @return {?}
     */
    stop(tag = 'general') {
        if (this.isLoadingSubjectsPool[tag] === undefined) {
            throw new Error(`Loading subject was not created for tag: ${tag}.`);
        }
        if (this.isLoadingSubjectsPool[tag].value > 0) {
            this.isLoadingSubjectsPool[tag].next(this.isLoadingSubjectsPool[tag].value - 1);
        }
    }
    /**
     * @param {?=} tag
     * @return {?}
     */
    destroy(tag = 'general') {
        if (this.isLoadingSubjectsPool[tag] !== undefined) {
            this.isLoadingSubjectsPool[tag].next(0);
            this.isLoadingSubjectsPool[tag].complete();
            delete this.isLoadingSubjectsPool[tag];
        }
    }
    /**
     * @private
     * @param {?} tag
     * @return {?}
     */
    checkAndInitIsLoadingSubject(tag) {
        if (this.isLoadingSubjectsPool[tag] === undefined) {
            this.isLoadingSubjectsPool[tag] = new BehaviorSubject(0);
        }
    }
}
MwLoadingService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root',
            },] }
];
/** @nocollapse */ MwLoadingService.ngInjectableDef = i0.ɵɵdefineInjectable({ factory: function MwLoadingService_Factory() { return new MwLoadingService(); }, token: MwLoadingService, providedIn: "root" });
if (false) {
    /**
     * @type {?}
     * @private
     */
    MwLoadingService.prototype.isLoadingSubjectsPool;
}
