// @mw-angular/core/esm2015/lib/services/mw-meta.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { Title } from '@angular/platform-browser';
import { BehaviorSubject } from 'rxjs';
import { MwEnvironmentService } from './mw-environment.service';
import * as i0 from "@angular/core";
import * as i1 from "./mw-environment.service";
import * as i2 from "@angular/platform-browser";
export class MwMetaService {
    /**
     * @param {?} mwEnvironmentService
     * @param {?} title
     */
    constructor(mwEnvironmentService, title) {
        this.mwEnvironmentService = mwEnvironmentService;
        this.title = title;
        this.envTitlePrefix = 'titlePrefix';
        this.envTitlePostfix = 'titlePostfix';
        this.pageTitleSubject = new BehaviorSubject('');
    }
    /**
     * @param {?} title
     * @return {?}
     */
    setTitle(title) {
        /** @type {?} */
        const prefix = this.mwEnvironmentService.getValue(this.envTitlePrefix, '');
        /** @type {?} */
        const postfix = this.mwEnvironmentService.getValue(this.envTitlePostfix, '');
        this.title.setTitle(`${prefix}${title}${postfix}`);
        this.pageTitleSubject.next(title);
    }
    /**
     * @return {?}
     */
    getPageTitle() {
        return this.pageTitleSubject.asObservable();
    }
}
MwMetaService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root',
            },] }
];
/** @nocollapse */
MwMetaService.ctorParameters = () => [
    { type: MwEnvironmentService },
    { type: Title }
];
/** @nocollapse */ MwMetaService.ngInjectableDef = i0.ɵɵdefineInjectable({ factory: function MwMetaService_Factory() { return new MwMetaService(i0.ɵɵinject(i1.MwEnvironmentService), i0.ɵɵinject(i2.Title)); }, token: MwMetaService, providedIn: "root" });
if (false) {
    /**
     * @type {?}
     * @private
     */
    MwMetaService.prototype.envTitlePrefix;
    /**
     * @type {?}
     * @private
     */
    MwMetaService.prototype.envTitlePostfix;
    /**
     * @type {?}
     * @private
     */
    MwMetaService.prototype.pageTitleSubject;
    /**
     * @type {?}
     * @private
     */
    MwMetaService.prototype.mwEnvironmentService;
    /**
     * @type {?}
     * @private
     */
    MwMetaService.prototype.title;
}
