// @mw-angular/core/esm2015/lib/services/mw-platform.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { isPlatformBrowser, isPlatformServer } from '@angular/common';
import { Inject, Injectable, PLATFORM_ID } from '@angular/core';
import * as i0 from "@angular/core";
export class MwPlatformService {
    /**
     * @param {?} platformId
     */
    constructor(platformId) {
        this.platformId = platformId;
    }
    /**
     * @return {?}
     */
    isBrowser() {
        return isPlatformBrowser(this.platformId);
    }
    /**
     * @return {?}
     */
    isServer() {
        return isPlatformServer(this.platformId);
    }
}
MwPlatformService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root',
            },] }
];
/** @nocollapse */
MwPlatformService.ctorParameters = () => [
    { type: undefined, decorators: [{ type: Inject, args: [PLATFORM_ID,] }] }
];
/** @nocollapse */ MwPlatformService.ngInjectableDef = i0.ɵɵdefineInjectable({ factory: function MwPlatformService_Factory() { return new MwPlatformService(i0.ɵɵinject(i0.PLATFORM_ID)); }, token: MwPlatformService, providedIn: "root" });
if (false) {
    /**
     * @type {?}
     * @private
     */
    MwPlatformService.prototype.platformId;
}
