// @mw-angular/core/esm2015/lib/services/mw-request.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { DOCUMENT } from '@angular/common';
import { Inject, Injectable, Optional } from '@angular/core';
import { REQUEST } from '@nguniversal/express-engine/tokens';
import { MwPlatformService } from './mw-platform.service';
import * as i0 from "@angular/core";
import * as i1 from "./mw-platform.service";
import * as i2 from "@angular/common";
import * as i3 from "@nguniversal/express-engine/tokens";
export class MwRequestService {
    /**
     * @param {?} mwPlatformService
     * @param {?} document
     * @param {?} request
     */
    constructor(mwPlatformService, document, request) {
        this.mwPlatformService = mwPlatformService;
        this.document = document;
        this.request = request;
    }
    /**
     * @return {?}
     */
    hasRequest() {
        return (this.mwPlatformService.isServer() && this.request !== null) || this.mwPlatformService.isBrowser();
    }
    /**
     * Example localhost:5000 | www.domain.com
     * @return {?}
     */
    getHost() {
        /** @type {?} */
        let host;
        if (this.mwPlatformService.isServer()) {
            if (this.request === null) {
                throw new Error('Request object in not set.');
            }
            host = this.request.get('host');
        }
        else if (this.mwPlatformService.isBrowser()) {
            host = this.document.location.host;
        }
        else {
            throw new Error('Unknown platform.');
        }
        if (typeof host === 'undefined') {
            throw new Error('Host is undefined.');
        }
        return host;
    }
    /**
     * Example http://localhost:5000 | https://www.domain.com
     * @return {?}
     */
    getLocationOrigin() {
        /** @type {?} */
        let locationOrigin;
        if (this.mwPlatformService.isServer()) {
            if (this.request === null) {
                throw new Error('Request object in not set.');
            }
            /** @type {?} */
            const host = this.request.get('host');
            if (typeof host === 'undefined') {
                throw new Error('Host is undefined.');
            }
            /** @type {?} */
            const protocol = this.request.protocol;
            locationOrigin = `${protocol}://${host}`;
        }
        else if (this.mwPlatformService.isBrowser()) {
            locationOrigin = this.document.location.origin;
        }
        else {
            throw new Error('Unknown platform.');
        }
        return locationOrigin;
    }
    /**
     * Example: if url is https://www.domain.com/en/about?qp=1 method returns '/en/about'
     * @return {?}
     */
    getLocationPathname() {
        /** @type {?} */
        let locationPathname;
        if (this.mwPlatformService.isServer()) {
            if (this.request === null) {
                throw new Error('Request object in not set.');
            }
            locationPathname = this.request.path;
        }
        else if (this.mwPlatformService.isBrowser()) {
            locationPathname = this.document.location.pathname;
        }
        else {
            throw new Error('Unknown platform.');
        }
        return locationPathname;
    }
}
MwRequestService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root',
            },] }
];
/** @nocollapse */
MwRequestService.ctorParameters = () => [
    { type: MwPlatformService },
    { type: undefined, decorators: [{ type: Inject, args: [DOCUMENT,] }] },
    { type: undefined, decorators: [{ type: Optional }, { type: Inject, args: [REQUEST,] }] }
];
/** @nocollapse */ MwRequestService.ngInjectableDef = i0.ɵɵdefineInjectable({ factory: function MwRequestService_Factory() { return new MwRequestService(i0.ɵɵinject(i1.MwPlatformService), i0.ɵɵinject(i2.DOCUMENT), i0.ɵɵinject(i3.REQUEST, 8)); }, token: MwRequestService, providedIn: "root" });
if (false) {
    /**
     * @type {?}
     * @private
     */
    MwRequestService.prototype.mwPlatformService;
    /**
     * @type {?}
     * @private
     */
    MwRequestService.prototype.document;
    /**
     * @type {?}
     * @private
     */
    MwRequestService.prototype.request;
}
